src/quicklist.c

```diff
@@ -1254,7 +1263,12 @@ quicklistIter *quicklistGetIteratorAtIdx(const quicklist *quicklist,
 /* Release iterator.
  * If we still have a valid current node, then re-encode current node. */
 void quicklistReleaseIterator(quicklistIter *iter) {
-    if (iter->current)
+    /* When an iterator is unsafe, the user of the iterator may have
+     * performed operations against the quicklist while iterating,
+     * so we need to recompress the entire quicklist again. */
+    if (iter->safe)
+        __quicklistCompress(iter->quicklist, NULL);
```

Comment:
i don't like that too much.
correct me if i'm wrong, if we have an 1000 node quicklist.
and we started a safe iterator and stepped only on the first 5 nodes.
we'll still iterate over all 1000 nodes when we close the iterator?

Reply:
Let me think again if there is a better way.

Reply:
Use a different way to deal with recompress issues.
1. Remove `quicklistEntry` completely, and remove `listTypeEntry` at the same time
   The main reason for the recompress exception is that none of the public methods of the quicklist can guarantee that the quicklistIter is updated correctly.
   By deleting the quicklistEntry, all our operations against quicklist will be based on the `quicklistIter`, so that we can ensure that all operations will update the `quicklistIter->current` correctly.

2. Delete `quicklistIndex`.
    In any case, using `quicklistIndex` to get a quicklistEntry will result in the quicklistEntry not being recompressed, using quicklistGetIteratorAtIdx instead.